// @lugia/theme-config/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var theme_core_1 = require("@lugia/theme-core");
var React = tslib_1.__importStar(require("react"));
var Theme = /** @class */ (function (_super) {
    tslib_1.__extends(Theme, _super);
    function Theme(props, context) {
        var _this = _super.call(this, props) || this;
        _this.svThemeConfigTree = {};
        _this.context = { config: {}, svThemeConfigTree: {} };
        _this.updateTreeConfig(props, context);
        return _this;
    }
    Theme.prototype.componentWillReceiveProps = function (nextProps, context) {
        var nowContext = this.context;
        if (nextProps.config !== this.props.config ||
            nowContext.config !== context.config ||
            nowContext.svThemeConfigTree !== context.svThemeConfigTree) {
            this.updateTreeConfig(nextProps, context);
        }
    };
    Theme.prototype.updateTreeConfig = function (props, context) {
        var config = context.config, svThemeConfigTree = context.svThemeConfigTree;
        this.svThemeConfigTree = theme_core_1.getConfig(svThemeConfigTree, config, props.config);
    };
    Theme.prototype.getContextValue = function () {
        var props = this.props;
        var config = props.config;
        return {
            config: config,
            svThemeConfigTree: this.svThemeConfigTree,
        };
    };
    Theme.prototype.render = function () {
        var children = this.props.children;
        return (React.createElement(theme_core_1.ThemeContext.Provider, { value: this.getContextValue() }, children));
    };
    Theme.defaultProps = {
        config: {},
    };
    Theme.contextType = theme_core_1.ThemeContext;
    Theme.displayName = 'lugia_theme_wrap';
    return Theme;
}(React.Component));
exports.default = Theme;
